context-free.md

```diff
@@ -2,21 +2,221 @@
+# Internal or external?
+_Is the language implemented as an internal or external DSL? 
+Justify your answer._
+
+I argue that HQL is implemented as an external DSL.
+
+The book says that "an external DSL is a language separate from the main language 
+of the application it works in. 
+A script in an external DSL will usually be parsed by code in the host application 
+using text parsing techniques. Examples of external DSLs... include SQL."
+
+HQL seems to be an absttraction of SQL;
+HQL queries are translated/parsed by Hibernate into conventional SQL queries, 
+SQL being an external DSL itself.
+
+
+# Host language
+_What language(s) was (were) used to implement the DSL?_
+
+Hibernate is an object-relational mapping framework for the Java language. 
+It provides a framework for mapping an object-oriented domain model to a relational database. 
+Hibernate solves object-relational impedance mismatch problems by replacing direct, 
+persistent database accesses with high-level object handling functions.
+
+
+# Benefits
+_Identify one potential benefit of the DSL: how is a programmer's life or a 
+company's bottom line made easier by the existence of this language?_
+
+The following are several benefits:
+
+1. Similar to SQL: HQL’s syntax is very similar to that of standard SQL. 
+Familiarity with SQL makes it easy to learn HQL. 
+
+2. Fully object-oriented:
+HQL does not use real names of table and columns - 
+it uses class and property names instead. 
+HQL can understand inheritance, polymorphism and association.
+
+3. Case-insensitive for keywords: 
+Like SQL, keywords in HQL are case-insensitive. 
+That means SELECT, select or Select are the same.
+
+4. Case-sensitive for Java classes and properties: 
+HQL considers case-sensitive names for Java classes 
+and their properties, meaning Person and person are two different objects.
+
+5. Aides coding: 
+HQL helps users conduct more thorough data analysis, 
+reduces development time, and ensures more accurate coding.
+
+6. Hides sensitive data:
+Since a class defines only the data it needs to be concerned with, 
+when an instance of that class (an object) is run, 
+the code will not be able to accidentally access other program data. 
+This characteristic of data hiding provides greater system security 
+and avoids unintentional data corruption.
+
+7. Can be understood by people who cannot code in other languages:
+Anyone can quickly learn to write in HQL, 
+which is a helpful abstraction of SQL and general database manipulation,
+instead of having to learning a more complex and low-level general-purpose language.
+
+8. Simplifies details of database management that one might not need to know:
+Database management is complicated, and has been considerably simplified 
+by the SQL programming language. SQL is responsible for querying and editing 
+information stored in a certain database management system, and HQL makes it even 
+easier to manipulate data in the form of class names and property names. 
```

Comment:
It seems like there must be a major difference beyond being able to use class and property names - most languages/frameworks provide a lot of these features (and more) without using a new (external) DSL. C# comes to mind, with both LINQ and Entity Framework (although arguably both are DSLs...).